### ISSUE TITLE
Commit 1df074af1dc6b32a4f5b704f3f31c552cb11c9a8 made pkgcache_test flaky

### ISSUE THREAD

**Issue by bsilver8192** [bazelbuild/bazel#1154]
ci.bazel.io has [a few failures](http://ci.bazel.io/job/Bazel/475/), and I am seeing it fail most of the time (failed 16 out of 20 times when I tried it). Stack trace is consistent:
```
com.google.devtools.build.lib.pkgcache.TargetPatternEvaluatorTest.testSubdirectoryCircularSymlinkNoKeepGoing

Error Message
com.google.devtools.build.lib.packages.BuildFileNotFoundException: no such package 'parent/child': Symlink cycle detected while trying to find BUILD file /workspace/parent/child/BUILD

Stacktrace
java.lang.IllegalStateException: com.google.devtools.build.lib.packages.BuildFileNotFoundException: no such package 'parent/child': Symlink cycle detected while trying to find BUILD file /workspace/parent/child/BUILD
	at com.google.devtools.build.lib.skyframe.RecursiveDirectoryTraversalFunction$1.apply(RecursiveDirectoryTraversalFunction.java:124)
	at com.google.devtools.build.lib.skyframe.RecursiveDirectoryTraversalFunction$1.apply(RecursiveDirectoryTraversalFunction.java:118)
	at com.google.common.collect.Maps$7.transformEntry(Maps.java:1868)
	at com.google.common.collect.Maps$10.getValue(Maps.java:1913)
	at com.google.common.collect.Maps$EntryFunction$2.apply(Maps.java:100)
	at com.google.common.collect.Maps$EntryFunction$2.apply(Maps.java:96)
	at com.google.common.collect.Iterators$8.transform(Iterators.java:817)
	at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:48)
	at com.google.devtools.build.lib.skyframe.RecursivePkgFunction$MyTraversalFunction.aggregateWithSubdirectorySkyValues(RecursivePkgFunction.java:81)
	at com.google.devtools.build.lib.skyframe.RecursivePkgFunction$MyTraversalFunction.aggregateWithSubdirectorySkyValues(RecursivePkgFunction.java:53)
	at com.google.devtools.build.lib.skyframe.RecursiveDirectoryTraversalFunction.visitDirectory(RecursiveDirectoryTraversalFunction.java:321)
	at com.google.devtools.build.lib.skyframe.RecursivePkgFunction.compute(RecursivePkgFunction.java:50)
	at com.google.devtools.build.skyframe.ParallelEvaluator.bubbleErrorUp(ParallelEvaluator.java:1493)
	at com.google.devtools.build.skyframe.ParallelEvaluator.waitForCompletionAndConstructResult(ParallelEvaluator.java:1342)
	at com.google.devtools.build.skyframe.ParallelEvaluator.eval(ParallelEvaluator.java:1311)
	at com.google.devtools.build.skyframe.ParallelEvaluator.eval(ParallelEvaluator.java:1272)
	at com.google.devtools.build.skyframe.InMemoryMemoizingEvaluator.evaluate(InMemoryMemoizingEvaluator.java:189)
	at com.google.devtools.build.skyframe.SequentialBuildDriver.evaluate(SequentialBuildDriver.java:39)
	at com.google.devtools.build.lib.skyframe.SkyframeExecutor.targetPatterns(SkyframeExecutor.java:1106)
	at com.google.devtools.build.lib.skyframe.SkyframeTargetPatternEvaluator.parseTargetPatternKeys(SkyframeTargetPatternEvaluator.java:134)
	at com.google.devtools.build.lib.skyframe.SkyframeTargetPatternEvaluator.parseTargetPatternList(SkyframeTargetPatternEvaluator.java:117)
	at com.google.devtools.build.lib.skyframe.SkyframeTargetPatternEvaluator.parseTargetPatternList(SkyframeTargetPatternEvaluator.java:56)
	at com.google.devtools.build.lib.pkgcache.AbstractTargetPatternEvaluatorTest.parseTargetPatternList(AbstractTargetPatternEvaluatorTest.java:60)
	at com.google.devtools.build.lib.pkgcache.AbstractTargetPatternEvaluatorTest.parseTargetPatternList(AbstractTargetPatternEvaluatorTest.java:52)
	at com.google.devtools.build.lib.pkgcache.TargetPatternEvaluatorTest.parseList(TargetPatternEvaluatorTest.java:157)
	at com.google.devtools.build.lib.pkgcache.TargetPatternEvaluatorTest.testSubdirectoryCircularSymlinkNoKeepGoing(TargetPatternEvaluatorTest.java:672)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at com.google.testing.junit.runner.junit4.CancellableRequestFactory$CancellableRunner.run(CancellableRequestFactory.java:84)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
	at com.google.testing.junit.runner.junit4.JUnit4Runner.run(JUnit4Runner.java:103)
	at com.google.testing.junit.runner.BazelTestRunner.runTestsInSuite(BazelTestRunner.java:149)
	at com.google.testing.junit.runner.BazelTestRunner.main(BazelTestRunner.java:90)
Caused by: com.google.devtools.build.lib.packages.BuildFileNotFoundException: no such package 'parent/child': Symlink cycle detected while trying to find BUILD file /workspace/parent/child/BUILD
	at com.google.devtools.build.lib.skyframe.PackageLookupFunction.getFileValue(PackageLookupFunction.java:126)
	at com.google.devtools.build.lib.skyframe.PackageLookupFunction.getPackageLookupValue(PackageLookupFunction.java:140)
	at com.google.devtools.build.lib.skyframe.PackageLookupFunction.compute(PackageLookupFunction.java:92)
	... 47 more
```

@janakdr: Looks like the wrong exception gets thrown sometimes with that recent change of yours.

**Comment by janakdr** [comment]
See https://bazel-review.googlesource.com/#/c/3392/

**Comment by damienmg** [comment]
Assigning to @janakdr as he already sent a fix. This should have been triaged days ago...

**Comment by janakdr** [comment]
Fixed by fbafe83b65965277e8c8c53b3f4b978b5507e021
